### PR TITLE
[#5384] Added OAuth support to Gravitino CLI

### PR DESCRIPTION
**Pull request by justinmclean**
### What changes were proposed in this pull request?

 Added OAuth support to Gravitino CLI, Note this is supported by the configuration files.

### Why are the changes needed?

To support OAuth in the Gravitino CLI.

Fix: #5384

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Tested locally.

### CODE REVIEW

File: clients/cli/src/main/java/org/apache/gravitino/cli/GravitinoCommandLine.java [hunk 1]
```diff
@@ -337,167 +346,159 @@ private void handleSchemaCommand() {
...
-        new TableDistribution(url, ignore, metalake, catalog, schema, table).handle();
-      } else if (line.hasOption(GravitinoOptions.Partition)) {
-        new TablePartition(url, ignore, metalake, catalog, schema, table).handle();
+        new TableDistribution(url, ignore, auth, userName, metalake, catalog, schema, table)
+            .handle();
+      } else if (line.hasOption(GravitinoOptions.PARTITION)) {
+        new TablePartition(url, ignore, auth, userName, metalake, catalog, schema, table).handle();
       } else {
-        new TableDetails(url, ignore, metalake, catalog, schema, table).handle();
+        new TableDetails(url, ignore, auth, userName, metalake, catalog, schema, table).handle();
```
Comment (tengqm) [inline clients/cli/src/main/java/org/apache/gravitino/cli/GravitinoCommandLine.java:374]: A little bit concerned that the list of params is getting longer and longer.

File: clients/cli/src/main/java/org/apache/gravitino/cli/GravitinoCommandLine.java [hunk 2]
```diff
@@ -554,4 +557,41 @@ public String getUrl() {
...
+
+    // If specified on the command line use that
+    if (line.hasOption(GravitinoOptions.SIMPLE)) {
+      return GravitinoOptions.SIMPLE;
+    }
+
+    // Cache the Gravitino URL environment variable
+    if (authEnv == null && !authSet) {
+      authEnv = System.getenv("GRAVITINO_AUTH");
+      authSet = true;
```
Comment (tengqm) [inline clients/cli/src/main/java/org/apache/gravitino/cli/GravitinoCommandLine.java:578]: isn't this decision made too early?
Move it to line before 583?

File: clients/cli/src/main/java/org/apache/gravitino/cli/commands/AuditCommand.java [hunk 3]
```diff
@@ -25,9 +25,11 @@ public abstract class AuditCommand extends Command {
   /**
    * @param url The URL of the Gravitino server.
    * @param ignoreVersions If true don't check the client/server versions match.
+   * @param authentication Authentication type i.e. "simple"
+   * @param userName User name for simple authentication.
    */
-  public AuditCommand(String url, boolean ignoreVersions) {
-    super(url, ignoreVersions);
+  public AuditCommand(String url, boolean ignoreVersions, String authentication, String userName) {
+    super(url, ignoreVersions, authentication, userName);
```
Comment (tengqm) [inline clients/cli/src/main/java/org/apache/gravitino/cli/commands/AuditCommand.java:32]: Same as above, please consider a thorough check.

File: docs/cli.md [hunk 4]
```diff
@@ -522,51 +533,15 @@ gcli tag update --tag tagA --rename newTag
...
-```
-
-#### Create a role
-
-```bash
-gcli role create --role admin
-```
-
-#### Delete a role
+gcli <normal command> --simple
```
Comment (tengqm) [inline docs/cli.md:541]: ?
```suggestion
gcli <normal command> --auth simple
```

File: docs/cli.md [hunk 5]
```diff
@@ -522,51 +533,15 @@ gcli tag update --tag tagA --rename newTag
...
-
-```bash
-gcli group grant --group groupA --role admin
-```
+#### Simple authentication with user name
 
-#### Remove a role from a group
 ```bash
-gcli group revoke  --group groupA --role admin
+gcli <normal command> --simple --login userName
```
Comment (tengqm) [inline docs/cli.md:546]: ?
```suggestion
gcli <normal command> --auth simple --login userName
```

File: clients/cli/src/main/java/org/apache/gravitino/cli/GravitinoCommandLine.java (hunk 1 above)
Comment (justinmclean) [reply clients/cli/src/main/java/org/apache/gravitino/cli/GravitinoCommandLine.java:374]: This is in another PR that has not been merged and is unrelated to this PR.

File: clients/cli/src/main/java/org/apache/gravitino/cli/GravitinoOptions.java [hunk 6]
```diff
@@ -42,10 +42,12 @@ public class GravitinoOptions {
...
   public static final String ROLE = "role";
   public static final String AUDIT = "audit";
-  public static final String INDEX = "index";
   public static final String FORCE = "force";
+  public static final String SIMPLE = "simple";
+  public static final String LOGIN = "login";
+  public static final String INDEX = "index";
   public static final String DISTRIBUTION = "distribution";
-  public static final String Partition = "partition";
+  public static final String PARTITION = "partition";
```
[the remark this answers is not in the document]
Comment (justinmclean) [reply clients/cli/src/main/java/org/apache/gravitino/cli/GravitinoOptions.java:50]: How would you suggest refactoring it?

File: clients/cli/src/main/java/org/apache/gravitino/cli/commands/AuditCommand.java (hunk 3 above)
Comment (justinmclean) [reply clients/cli/src/main/java/org/apache/gravitino/cli/commands/AuditCommand.java:32]: same again

File: docs/cli.md (hunk 4 above)
Comment (justinmclean) [reply docs/cli.md:541]: Not related to this PR. However, we could possibly do it that way, but as this is the only authentication type that will be directly supported via the command line as others will be supported via the configuration file.

File: docs/cli.md (hunk 5 above)
Comment (justinmclean) [reply docs/cli.md:546]: seee above

File: clients/cli/src/main/java/org/apache/gravitino/cli/GravitinoOptions.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (tengqm) [reply clients/cli/src/main/java/org/apache/gravitino/cli/GravitinoOptions.java:50]: for example, group the options/operations by user role;
for another example, sort the list in alphabet order.

**Comment by justinmclean** [comment]
@shaofengshi this is ready to review

File: clients/cli/src/main/java/org/apache/gravitino/cli/commands/Command.java [hunk 7]
```diff
@@ -115,12 +136,26 @@ protected GravitinoAdminClient buildAdminClient() {
...
     }
     if (authentication != null) {
-      if (authentication.equals("simple")) {
+      if (authentication.equals(SIMPLE_AUTH)) {
         if (userName != null && !userName.isEmpty()) {
           client = client.withSimpleAuth(userName);
         } else {
           client = client.withSimpleAuth();
         }
+      } else if (authentication.equals(OAUTH_AUTH)) {
```
Comment (shaofengshi) [inline clients/cli/src/main/java/org/apache/gravitino/cli/commands/Command.java:145]: This peace of code is duplicated with the buildClient() method; is there a way to reuse?

File: clients/cli/src/main/java/org/apache/gravitino/cli/GravitinoConfig.java [hunk 8]
```diff
@@ -94,6 +95,15 @@ public void read() {
     if (prop.containsKey(authKey)) {
       authentication = prop.getProperty(authKey);
     }
+
+    if (authKey.equals("oauth")) {
```
Comment (shaofengshi) [inline clients/cli/src/main/java/org/apache/gravitino/cli/GravitinoConfig.java:99]: Can this use "OAUTH_AUTH" to replace "oauth"?

File: clients/cli/src/main/java/org/apache/gravitino/cli/commands/Command.java [hunk 9]
```diff
@@ -21,23 +21,30 @@
...
-  public static String OUTPUT_FORMAT_TABLE = "table";
-  public static String OUTPUT_FORMAT_PLAIN = "plain";
+  public static final String OUTPUT_FORMAT_TABLE = "table";
+  public static final String OUTPUT_FORMAT_PLAIN = "plain";
 
   protected static String authentication = null;
   protected static String userName = null;
 
+  private static final String SIMPLE_AUTH = "simple";
+  private static final String OAUTH_AUTH = "oauth";
```
Comment (shaofengshi) [inline clients/cli/src/main/java/org/apache/gravitino/cli/commands/Command.java:42]: This is very minor: about the naming convension, will "AUTH_SIMPLE" /  "AUTH_OAUTH" be better than "SIMPLE_AUTH" / "OAUTH_AUTH"? As in the above, I see "OUTPUT_FORMAT_TABLE" and "OUTPUT_FORMAT_PLAIN", they share the same prefix.

File: clients/cli/src/main/java/org/apache/gravitino/cli/commands/Command.java (hunk 7 above)
Comment (justinmclean) [reply clients/cli/src/main/java/org/apache/gravitino/cli/commands/Command.java:145]: There was no way I coudl find to reuse it I did try. One part of the code uses GravitinoAdminClient and the other GravitinoClient, which are not interchangeable when using the Builder.

File: clients/cli/src/main/java/org/apache/gravitino/cli/GravitinoConfig.java (hunk 8 above)
Comment (justinmclean) [reply clients/cli/src/main/java/org/apache/gravitino/cli/GravitinoConfig.java:99]: I didn't want this code to depend on Command as OAUTH_AUTH is private. I could extract to another class if you want.

File: clients/cli/src/main/java/org/apache/gravitino/cli/commands/Command.java (hunk 9 above)
Comment (justinmclean) [reply clients/cli/src/main/java/org/apache/gravitino/cli/commands/Command.java:42]: I went this was as other code uses "SIMPLE_AUTH_TYPE" I don't mind either way.

File: clients/cli/src/main/java/org/apache/gravitino/cli/commands/Command.java (hunk 7 above)
Comment (justinmclean) [reply clients/cli/src/main/java/org/apache/gravitino/cli/commands/Command.java:145]: If you extract the common code and try using it from both methods, you get errors with converting to/from Builder<GravitinoClient> and Builder<GravitinoAdminClient>.

Comment (justinmclean) [reply clients/cli/src/main/java/org/apache/gravitino/cli/commands/Command.java:145]: Looks like you can work around that by using `Builder<GravitinoClientBase>`, I thought it didn't have all the methods.

Comment (justinmclean) [reply clients/cli/src/main/java/org/apache/gravitino/cli/commands/Command.java:145]: Nope that doesn't work either, as you get "Builder<GravitinoClient> cannot be converted to Builder<GravitinoClientBase>"

Comment (justinmclean) [reply clients/cli/src/main/java/org/apache/gravitino/cli/commands/Command.java:130]: finally worked out a way but it's not pretty.

File: clients/cli/src/main/java/org/apache/gravitino/cli/GravitinoConfig.java (hunk 8 above)
Comment (shaofengshi) [reply clients/cli/src/main/java/org/apache/gravitino/cli/GravitinoConfig.java:99]: Never mind, it is just a minor issue.

File: clients/cli/src/main/java/org/apache/gravitino/cli/commands/Command.java (hunk 9 above)
Comment (shaofengshi) [reply clients/cli/src/main/java/org/apache/gravitino/cli/commands/Command.java:43]: okay

**Review by shaofengshi — APPROVED** [review]
LGTM

**Pull request merged**